### PR TITLE
DROOLS-5333 : Allow to customize rule names in guided decision table editor

### PR DESCRIPTION
**Pull request by Rikkola**
https://issues.redhat.com/browse/DROOLS-5333

https://github.com/kiegroup/drools/pull/2904
https://github.com/kiegroup/drools-wb/pull/1370

### CODE REVIEW

**Comment by Rikkola** [comment]
@jomarko Just a heads up this PR is coming. I added +1 to about 2000 lines, but looks like the PRs are still failing. I will reping you when all this works.

**Comment by Rikkola** [comment]
Jenkins please retest this

**Comment by jomarko** [comment]
@Rikkola please have a look also on sonar report
- using diamond operator
- using isEmpty() instead if size() > 0
- do nto print stack trace in exception handling
use lambda instead of anonymous boolean callback

**Comment by Rikkola** [comment]
@jomarko I'm opening this for a proper review. I'm hoping this can get merged before #1365
 simply due to the fact that this touched everything and overlaps with it.

**Comment by Rikkola** [comment]
Weird the ShowRuleNameOptionPresenterTest works in Idea. I'll make a fix of some sort.

**Comment by jomarko** [comment]
Manual review is [here](https://issues.redhat.com/browse/DROOLS-5333?focusedCommentId=14119472&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14119472).

File: drools-wb-screens/drools-wb-dtable-xls-editor/drools-wb-dtable-xls-editor-backend/src/test/java/org/drools/workbench/screens/dtablexls/backend/server/conversion/GuidedDecisionTableGeneratorListenerTest.java [hunk 1]
```diff
@@ -3558,9 +3578,9 @@ private void checkConversionWithBigMaths(final ConversionResult result,
         //Check data
         assertEquals(2,
                      dtable.getData().size());
-        assertTrue(isRowEquivalent(new Object[]{1, "Hello World", "Message.HELLO", false, "Goodbye cruel world", "Message.GOODBYE", expectedDataValueRow0.get(), "m"},
+        assertTrue(isRowEquivalent(new Object[]{1, "", "Hello World", "Message.HELLO", false, "Goodbye cruel world", "Message.GOODBYE", expectedDataValueRow0.get(), "m"},
                                    dtable.getData().get(0)));
-        assertTrue(isRowEquivalent(new Object[]{2, "Goodbye", "Message.GOODBYE", false, "", "", expectedDataValueRow1.get(), "m"},
+        assertTrue(isRowEquivalent(new Object[]{2, "", "Goodbye", "Message.GOODBYE", false, "", "", expectedDataValueRow1.get(), "m"},
```
[the remark this answers is not in the document]
Comment (Rikkola) [reply drools-wb-screens/drools-wb-dtable-xls-editor/drools-wb-dtable-xls-editor-backend/src/test/java/org/drools/workbench/screens/dtablexls/backend/server/conversion/GuidedDecisionTableGeneratorListenerTest.java:3583]: Well spotted. The XLS tables let you define a rule name per column and this is missing it. This means I will have to also add the support for the GDST to XLS conversion once this is. For this workload I would propose that we make a ticket and I'll fix it for both ways once the two depending PRs are in.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/editor/page/ShowRuleNameOptionView.html [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<div>
+    <div class="rulename-wrapper">
+
+        <span class="rulename-wrapper-label" data-i18n-key="RuleName"></span>
+        <input type="checkbox" data-field="ruleNameCheckBox" id="kieScenarioSettingsStateless">
```
[the remark this answers is not in the document]
Comment (Rikkola) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/editor/page/ShowRuleNameOptionView.html:21]: Almost like I copy pasted this code from somewhere and forgot to remove this and change the license year.

File: drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/DTableUpdateManager.java [hunk 3]
```diff
@@ -70,8 +68,9 @@ public void update(final GuidedDecisionTable52 model,
         final Set<Integer> canBeUpdated = new HashSet<>();
 
         for (final Coordinate coordinate : coordinates) {
-            if (coordinate.getCol() != ROW_NUMBER_COLUMN
-                    && coordinate.getCol() != DESCRIPTION_COLUMN) {
+            if (coordinate.getCol() != GuidedDecisionTable52.RULE_NUMBER_INDEX
+                    && coordinate.getCol() != GuidedDecisionTable52.RULE_NAME_COLUMN_INDEX
+                    && coordinate.getCol() != GuidedDecisionTable52.RULE_DESCRIPTION_INDEX) {
```
[the remark this answers is not in the document]
Comment (Rikkola) [reply drools-wb-services/drools-wb-verifier/drools-wb-verifier-guided-decision-table-adapter/src/main/java/org/drools/workbench/services/verifier/plugin/client/DTableUpdateManager.java:73]: This has to change if anything is added before description anyway, so I guess greater than desc col is cleaner.

**Comment by Rikkola** [comment]
@jomarko I suspect the rule name column info was not saved since the presenter for the checkbox panel was not set. I actually have no idea how it worked for me. Unless my GWT generation was out of sync for some reason. Should work now.

**Comment by Rikkola** [comment]
Fixing the tests.

**Comment by Rikkola** [comment]
Tests fixed.

File: drools-wb-screens/drools-wb-dtable-xls-editor/drools-wb-dtable-xls-editor-backend/src/test/java/org/drools/workbench/screens/dtablexls/backend/server/conversion/GuidedDecisionTableGeneratorListenerTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jomarko) [reply drools-wb-screens/drools-wb-dtable-xls-editor/drools-wb-dtable-xls-editor-backend/src/test/java/org/drools/workbench/screens/dtablexls/backend/server/conversion/GuidedDecisionTableGeneratorListenerTest.java:3583]: https://issues.redhat.com/browse/DROOLS-5387

**Comment by jomarko** [comment]
@Rikkola , is there any chance we will include my test? https://github.com/Rikkola/drools-wb/pull/8/commits/506946a358a642c43cf30c472dcfef0456d3fe34

**Comment by Rikkola** [comment]
@jomarko yep we can.

**Comment by Rikkola** [comment]
@jomarko I fixed " for rule names and the pinn info location, but the last issue has been there for a while.

**Comment by Rikkola** [comment]
I'll get the sonar bugs fixed. It is unlikely we ever hit them, but better to be safe.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/test/java/org/drools/workbench/screens/guided/dtable/backend/server/GuidedDecisionTableSourceServiceTest.java [hunk 4]
```diff
@@ -243,6 +243,27 @@ public void testOtherwiseTwoTimes() throws Exception {
...
+        model.setData(data);
+
+        String source = service.getSource(path,
+
+                                          model);
+
+        assertTrue("Expected custom rule name to be present",
+
+                   source.contains("rule \"John is important\"\n"));
+    }
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-backend/src/test/java/org/drools/workbench/screens/guided/dtable/backend/server/GuidedDecisionTableSourceServiceTest.java:265]: Could we please eliminate count of empty new lines?

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/dom/textbox/TextBoxRuleNameSingletonDOMElementFactory.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2015 Red Hat, Inc. and/or its affiliates.
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/dom/textbox/TextBoxRuleNameSingletonDOMElementFactory.java:2]: 2020

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/dom/textbox/TextBoxRuleNameSingletonDOMElementFactory.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        final RuleNameTextBox widget = new RuleNameTextBox();
+        widget.addKeyDownHandler((e) -> e.stopPropagation());
+        widget.addMouseDownHandler((e) -> e.stopPropagation());
+        return widget;
+    }
+
+    @Override
+    public String convert(final String value) {
+        return value;
+    }
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/dom/textbox/TextBoxRuleNameSingletonDOMElementFactory.java:45]: We do not need to override this, it is same as in parent.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/RuleNameColumnConverter.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2011 Red Hat, Inc. and/or its affiliates.
```
[the remark this answers is not in the document]
Comment (jomarko) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/RuleNameColumnConverter.java:2]: 2020 please

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/RuleNameColumnConverter.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+                                                         access,
+                                                         gridWidget);
+        uiColumn.setMovable(false);
+        uiColumn.setFloatable(true);
+        uiColumn.setVisible(!column.isHideColumn());
+        uiColumn.setMinimumWidth(150.0);
+        uiColumn.setWidth(Math.max(column.getWidth(),
+                                   uiColumn.getMinimumWidth()));
+
+        return uiColumn;
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/RuleNameColumnConverter.java:57]: Isn't this equivalent to the below?
```java
        final GridColumn<?> uiColumn = newRuleNameColumn(makeHeaderMetaData(column),
                                                         Math.max(column.getWidth(),
                                                                  150),
                                                         true,
                                                         true,
                                                         access,
                                                         gridWidget);
        uiColumn.setMovable(false);
        uiColumn.setFloatable(true);
        uiColumn.setVisible(!column.isHideColumn());
        uiColumn.setMinimumWidth(150.0);
        return uiColumn;
```

We could declare 150 as class constant if answer is yes.

**Comment by jomarko** [comment]
During manual review found critical issue but not related to changes here. https://issues.redhat.com/browse/DROOLS-5407

Toni thank you for fixing pinned mode.Could you please also incorporate my comment about title for the added checkbox. It should be one-liner change and will help us to avoid adding some massive product documentation.

**Comment by jomarko** [comment]
@Rikkola here is the tooltip and we can merge I think. https://github.com/Rikkola/drools-wb/pull/9

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/RuleNameColumnConverter.java (hunk 8 above)
Comment (Rikkola) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/model/converters/column/impl/RuleNameColumnConverter.java]: I think it is. Did the change.

**Comment by Rikkola** [comment]
@kiegroup/gatekeepers Can I please get the 2 PRs in this bundle merged.

**Pull request merged**